### PR TITLE
ModBackground Rework

### PR DESCRIPTION
**Pull request by Lion8cake**
### What is the new feature?
This is a massive rework and change to ModSurface/UndergroundBackground. This contains both fixes to make modbackground easier to configure and edit as well as fixes to make this be on par with vanilla background drawing. 

### Why should this be part of tModLoader?
Currently ModSurfaceBackground has almost no customisability and when it does its only for the Close background stuff. Even the comment on the current close background texture hook mentions that whoever wrote the summary has no idea what half the parameters even do. Along with this, mod backgrounds are only limited to 3 layers (Far, Middle, Close) while vanilla has a maximum of 5.

This should make modding backgrounds easier and to allow those who want their backgrounds to have the same movement, styling, and height of vanilla bgs can do so.

Predraws have also been given to both the surface and underground background styles along allowing for extra drawing on top or in replacement of the mod background.

This also fixes World Globe changing forest backgrounds when in a mod biome.

Lastly this contains example mod additions to backgrounds to show how CloseMid and CloseFar work along with an extra part for the example mod underground transition between the lava and cavern layer.

### Are there alternative designs?
A completely rewrite of the mod background. I’m not advanced enough to rewrite the background from scratch so someone more advanced may want to completely rewrite modbackground to do what i have done in addition with some additional content.

### Sample usage for the new feature
Please see the following:
[ExampleMod/Content/Biomes/ExampleSurfaceBackgroundStyle.cs](https://github.com/tModLoader/tModLoader/compare/1.4.4...Lion8cake:tModLoader:1.4.4#diff-7b459f9331e86ab846c1f12d82f84748e45db630c20472cb51008e7555434e7d)
[ExampleMod/Content/Biomes/ExampleUndergroundBackgroundStyle.cs](https://github.com/tModLoader/tModLoader/compare/1.4.4...Lion8cake:tModLoader:1.4.4#diff-5076fa13d6f12119f825a52170fd054daa8c1d00ac647faf1cbf817b6eb98d00)

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs [hunk 1]
```diff
@@ -43,44 +48,87 @@ protected override sealed void Register()
 	/// <summary>
 	/// Allows you to modify the transparency of all background styles that exist. In general, you should move the index equal to this style's slot closer to 1, and all other indexes closer to 0. The transitionSpeed parameter is what you should add/subtract to each element of the fades parameter. See the ExampleMod for an example.
 	/// </summary>
-	public abstract void ModifyFarFades(float[] fades, float transitionSpeed);
+	public abstract void ModifyStyleFade(float[] fades, float transitionSpeed);
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs]: The original overloads of `ChooseXTexture` and `ModifyFarFades` must be kept present in form of `[Obsolete(error: true)]` methods, with `ModifyFarFades` converted to `virtual`. Refer to 1fc38455f564a9ac462219d467c1ae013002dfb9 and [this PR](https://github.com/tModLoader/tModLoader/commit/1fc38455f564a9ac462219d467c1ae013002dfb9) as a reference. The short summary of "the trick" here is that compile-time erroring methods can only be called from other obsolete methods, so a special non-erroring calling method must be made.

File: patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs [hunk 2]
```diff
@@ -43,44 +48,87 @@ protected override sealed void Register()
...
-	/// Allows you to determine which texture is drawn in the middle of the background.
+	/// Allows you to determine which texture is drawn in the middle of the background. This also lets you modify the scale, parallax, Y value and Looping Width.
 	/// </summary>
-	public virtual int ChooseMiddleTexture()
+	/// <param name="scale">The scale.</param>
+	/// <param name="parallax">The parallax value.</param>
+	/// <param name="topY">The Top Y level that the BG is drawn at</param>
+	/// <param name="loopWidth">The looping width that the texture loops at. Defaults to (Main.screenWidth / Main.bgWidthScaled + 2)</param>
+	/// <returns></returns>
+	public virtual int ChooseMiddleTexture(ref float scale, ref double parallax, ref int topY, ref int loopWidth)
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs]: I have a feeling that someone will want to expand these overloads again in the future, so I'd prefer if this were a single-parameter overload that uses a special struct, `BackgroundTextureParameters`?.

**Review by Mirsario — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs (hunk 2 above)
Comment (Mirsario) [reply patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs]: Boss bars have a `BossBarDrawParams` struct, so `BackgroundTextureParams`/`BackgroundLayerParams` for consistency.

File: patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs (hunk 1 above)
Comment (Lion8cake) [reply patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs]: I do think Fox has fixed this fully now, open this request changes if you don't think so.

**Comment by Lion8cake** [comment]
Outdated pull, new one can be found here: https://github.com/tModLoader/tModLoader/pull/4156
(This was my own fault it was closed accidentally and I can't open it up since I can't change which fork the PR is based off)

**Pull request closed without merging**